File: frontend/src/schemas/inboundtemplate.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// InboundTemplate CRUD schemas. v0.8.13 ships the
// admin surface for the inbound_templates table
// (migration 0021_inbound_templates.sql, the
// `internal/inboundtemplates/` package from
// PR #205). The schemas here mirror
// `backend/internal/inboundtemplates/validate.go`
// and the request shapes in
// `backend/internal/inboundtemplates/handler.go`.
//
// # Why protocol + params are required on Create
//
// The panel's storage of `params` is opaque
// (`map[string]any`, JSONB) — the sing-box
// provider is the authoritative per-protocol
// schema validator. But protocol itself must
// be one of the closed set; the Go-side
// ValidationError rejects anything outside
// the set with field=protocol. The form-side
// zod schema enforces the same set so the
// operator never sends a value the panel
// will reject.
//
// # Update semantics
//
// Partial update (the same PATCH semantics the
// v0.2 CRUD surfaces use). Every field is
// optional; the absence of a key means "leave
// unchanged". `protocol` IS patchable on its
// own (rare, but the Go Service supports it);
// the inbound-tables' `validateProtocol` rule
// still applies.

import { z } from 'zod'

import { protocolSchema } from './inbound'

// Description is a free-form note. The Go side
// caps it at 256 chars (matches the table column
// width); we enforce the same on the form so the
// operator sees the limit before submit.
const descriptionSchema = z
  .string()
  .max(256, 'Description must be 256 characters or fewer')
  .optional()

// params is opaque JSON. The form surfaces a
// textarea; the panel's store accepts whatever
// JSON the user pastes. Per-protocol shape
// validation lives in the sing-box provider
// (downstream of the panel).
const paramsSchema = z.record(z.unknown()).default({})

export const inboundTemplateCreateSchema = z.object({
  name: z.string().min(1).max(64),
  protocol: protocolSchema,
  params: paramsSchema,
  description: descriptionSchema,
})

export type InboundTemplateCreateInput = z.infer<typeof inboundTemplateCreateSchema>

export const inboundTemplateUpdateSchema = z.object({
  name: z.string().min(1).max(64).optional(),
  protocol: protocolSchema.optional(),
  params: paramsSchema.optional(),
  description: descriptionSchema,
})

export type InboundTemplateUpdateInput = z.infer<typeof inboundTemplateUpdateSchema>

// InboundTemplate is the full row shape the
// GET /api/v1/inbound-templates/{id} response
// returns. The same shape is used for each item
// in the list response. Mirrors the Go-side
// `InboundTemplate` struct one-to-one
// (camelCase after the json-tag normalisation
// the v0.2 wire-format fix introduced).
export interface InboundTemplate {
  id: string
  name: string
  protocol: z.infer<typeof protocolSchema>
  params: Record<string, unknown>
  description?: string | null
  createdAt: string
  updatedAt: string
}

// listInboundTemplatesResponse is the unwrapped
// list shape (the HTTP envelope has a
// `{templates: [...]}` wrapper; services strip
// it before returning to callers).
export interface InboundTemplateListResponse {
  templates: InboundTemplate[]
}

// The Create response returns the full row
// (with server-assigned id + timestamps), so
// callers can immediately show the new template
// in the table without a second GET.
export type InboundTemplateCreateResponse = InboundTemplate

File: backend/internal/inboundtemplates/inboundtemplate.go
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// Package inboundtemplates implements the panel-side CRUD for
// the `inbound_templates` table (migration
// 0021_inbound_templates.sql).
//
// # Model
//
// An InboundTemplate is a *named, reusable* protocol
// configuration that any number of `inbounds` rows on
// any node can reference via the new nullable FK
// `inbounds.template_id`. The template carries the
// protocol-specific configuration blob the sing-box
// provider's RenderConfig already understands
// (Reality keys, UUIDs, passwords, …) — the same
// shape as `inbounds.params`, but factored out so the
// operator does not paste the same JSON into every
// inbound.
//
// Templates are global (not per-node) — a VLESS-Reality
// template on 8443 with the same Reality keys can back
// twenty inbounds on twenty different nodes. The
// UNIQUE (name) constraint enforces one-name-per-panel
// (the templates are shared resources).
//
// # Relationship to inbounds.params
//
// v0.8.0-v0.8.12 every inbound carries its own
// `params` JSONB. v0.8.13+ inbounds can opt into a
// template: when `template_id` is non-NULL, the
// sing-box renderer reads `template.params` instead
// of `inbound.params`. The inline `params` field is
// kept on the inbound for backwards compatibility —
// existing inbounds without a template_id are
// untouched, and the renderer picks one or the other
// at apply time.
//
// The per-user credential from
// `user_inbound_credentials` is still layered on top
// in the multi-user render — the template is the
// "shared protocol config", the per-user row is the
// "per-user auth credential". They do not collide
// (the multi-user render replaces the template's
// `uuid` / `password` keys with the per-user value;
// the rest of the params flow through unchanged).
package inboundtemplates

import (
	"time"

	"github.com/google/uuid"
)

// Protocol is the protocol family of an
// InboundTemplate. The string values match the
// sing-box provider's per-protocol renderers
// (internal/cores/singbox/) and are pinned by the DB
// CHECK constraint in migration
// 0021_inbound_templates.sql.
//
// We import the inbounds.Protocol type for the
// closed-set guarantee (the templates share the same
// set of allowed protocols as the inbounds that
// reference them) but expose it as our own
// re-typed alias so callers don't need the inbounds
// import just to construct a template.
type Protocol string

// Protocol values. Closed set — any value outside
// this list is rejected at the Service boundary.
const (
	ProtocolVLESS       Protocol = "vless"
	ProtocolHysteria2   Protocol = "hysteria2"
	ProtocolShadowsocks Protocol = "shadowsocks"
	ProtocolTrojan      Protocol = "trojan"
)

// InboundTemplate is the panel-side view of one row
// in the `inbound_templates` table. Fields mirror
// the table one-to-one; we keep them as a Go struct
// rather than `map[string]any` so handlers can rely
// on the type at compile time.
type InboundTemplate struct {
	ID uuid.UUID `json:"id"`
	// Name is the operator's human-readable label.
	// The DB UNIQUE (name) constraint enforces
	// one-name-per-panel.
	Name string `json:"name"`
	// Protocol is the protocol family. Closed set
	// (see the Protocol constants above). An inbound
	// that references this template must carry the
	// same protocol value (the Service layer
	// enforces the constraint at the inbounds CRUD
	// boundary; the DB does not).
	Protocol Protocol `json:"protocol"`
	// Params is the protocol-specific configuration
	// blob. The Go side stores it as map[string]any
	// because the per-protocol schema is owned by
	// the sing-box provider and would otherwise
	// duplicate the renderer's parameters. The
	// service validator does not enforce a shape
	// here; the sing-box provider's RenderConfig is
	// the authoritative schema check.
	Params map[string]any `json:"params"`
	// Description is operator-supplied free-form
	// notes ("VLESS-Reality for the EU fleet, rotated
	// quarterly"). Optional. The admin UI surfaces
	// it as a placeholder; no operator workflow
	// reads it.
	Description string `json:"description,omitempty"`

	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

// IsValid reports whether the template carries the
// minimum data the Service layer requires to accept
// it. It is intentionally permissive — heavy
// validation (cross-entity, length, format) lives in
// `Service.Create` so it can return rich per-field
// errors. IsValid is the cheap pre-flight check used
// by the store to reject obviously broken inserts.
func (t *InboundTemplate) IsValid() bool {
	if t == nil || t.Name == "" {
		return false
	}
	if t.Protocol == "" {
		return false
	}
	return true
}
